man/pandoc-citeproc.1.md
---
title: pandoc-citeproc
section: 1
...

# NAME

pandoc-citeproc - filter to resolve citations in a pandoc document.

# SYNOPSIS

pandoc-citeproc [options] [file..]

# DESCRIPTION
The `pandoc-citeproc` executable has two modes, filter mode and convert mode.

## Filter mode

Run without options, it acts as a filter that takes a
JSON-encoded Pandoc document, formats citations and adds a
bibliography, and returns a JSON-encoded pandoc document.
Citations will be resolved, and, assuming there are bibliography
entries, a bibliography will be inserted into a Div element with id
`refs`. If no such Div exists, one will be created and appended to the
end of the document (unless the `suppress-bibliography` metadata field
is set to a true value). If the CSL style species hanging indent, the
class `hanging-indent` will be added to this Div.  If you wish the
bibliography to have a section header, put the section header at the end
of your document (or at the beginning of the Div with id `refs`, if
there is one).  (See the `pandoc` (1) man page under "Citations" for
details on how to encode citations in pandoc's markdown.)

To process citations with pandoc, call pandoc-citeproc as a filter:

    pandoc --filter pandoc-citeproc input.md -s -o output.html

pandoc-citeproc will look for the following metadata fields in
the input.  (For details on how to set metadata fields in
pandoc's markdown, see the `pandoc` (1) man page under
`--metadata`, `--metadata-file`, and `Extension: yaml_metadata_block`.)

`bibliography`
:   A path, or YAML list of paths, of bibliography
    files to use.  These may be in any of the formats supported by
    bibutils.

    Format            File extension
    ------------      --------------
    BibLaTeX          .bib
    BibTeX            .bibtex
    Copac             .copac
    CSL JSON          .json
    CSL YAML          .yaml
    EndNote           .enl
    EndNote XML       .xml
    ISI               .wos
    MEDLINE           .medline
    MODS              .mods
    NBIB              .nbib
    RIS               .ris

    Note that `.bib` can generally be used with both BibTeX and BibLaTeX
    files, but you can use `.bibtex` to force BibTeX.

`references`
:   A YAML list of references.  Each reference is a YAML
    object.  The format is essentially CSL JSON format.  Here is an example:

        - id: doe2006
          author:
            family: Doe
            given: [John, F.]
          title: Article
          page: 33-34
          issued:
            year: 2006
          type: article-journal
          volume: 6
          container-title: Journal of Generic Studies

    The contents of fields will be interpreted as markdown when
    appropriate:  so, for example, emphasis and strong emphasis can
    be used in title fields. Simple tex math will also be
    parsed and rendered appropriately.

`csl` or `citation-style`
:   Path or URL of a CSL style file.
    If the file is not found relative to the working directory,
    pandoc-citeproc will look in the `$HOME/.csl` directory (or
    `C:\Users\USERNAME\AppData\Roaming\csl` in Windows 7).  If this is left
    off, pandoc-citeproc will look for `$HOME/.pandoc/default.csl`,
    and if this is not present, it will use
    `chicago-author-date.csl`, looking first in `$HOME/.csl` and
    then in its own data files.

`link-citations`
:   If this has a true value, citations in author-date and numerical
    styles will be hyperlinked to their corresponding bibliography entries.
    The default is not to add hyperlinks.

`citation-abbreviations`
:   Path to a CSL abbreviations JSON file.  If the file is not found
    relative to the working directory, pandoc-citeproc will look in the
    `$HOME/.csl` directory (or `C:\Users\USERNAME\AppData\Roaming\csl`
    in Windows 7).  The format is described
    [here](http://citationstylist.org/2011/10/19/abbreviations-for-zotero-test-release).
    Abbreviations are only output if, in the `.csl` file, `form="short"` is set on the 
    element that renders the variable.

    Here is a short example:

        { "default": {
            "container-title": {
                    "Lloyd's Law Reports": "Lloyd's Rep",
                    "Estates Gazette": "EG",
                    "Scots Law Times": "SLT"
            }
          }
        }

`lang`
:    Locale to use in formatting citations.  If this is not set, the
     locale is taken from the `default-locale` attribute of the CSL
     file.  `en-US` is used if a locale is not specified in either the
     metadata or the CSL file.  (For backwards compatibility,
     the field `locale` can be used instead of `lang`, but this
     `lang` should be used going forward.)

`suppress-bibliography`
:    If this has a true value, the bibliography will be left off.
     Otherwise a bibliography will be inserted into each
     Div element with id `refs`. If there is no such Div,
     one will be created at the end of the document.

`reference-section-title`
:    If this has a value, a section header with this title will be
     added before the bibliography.  If `reference-section-title`
     is not specified and the document ends with a section header,
     this final header will be treated as the bibliography header.
     Note: `reference-section-title` has no effect when the bibliography
     is placed explicitly using a Div with id `refs`.

`notes-after-punctuation`
:    If true (the default), pandoc will put footnote citations
     after following punctuation.  For example, if the source
     contains `blah blah [@jones99].`, the result will look like
     `blah blah.[^1]`, with the note moved after the period and
     the space collapsed.  If false, the space will still be
     collapsed, but the footnote will not be moved after the
     punctuation.

The metadata must contain either `references` or `bibliography` or
both as a source of references.  `csl` and `citation-abbreviations`
are optional.  If `csl` is not provided, a default stylesheet
will be used (either `~/.pandoc/default.csl` or a version of
`chicago-author-date.csl`).

## Convert mode

If the option `--bib2yaml` or `--bib2json` is used, `pandoc-citeproc` will
not process citations; instead, it will treat its input (from stdin or
files) as a bibliography and convert it either to a pandoc YAML metadata
section, suitable for inclusion in a pandoc document (`--bib2yaml`), or
as a CSL JSON bibliography, suitable for import to zotero (`--bib2json`).

The `--format` option can be used to specify the bibliography format,
though when files are used, `pandoc-citeproc` can generally guess this
from the extension.

This mode supersedes the old `biblio2yaml` program.

## Raw content (pandoc-citeproc only)

To include raw content in a prefix, suffix, delimiter, or term,
surround it with these tags indicating the format:

    {{jats}}&lt;ref&gt;{{/jats}}

Without the tags, the string will be interpreted as a string
and escaped in the output, rather than being passed through raw.

This feature allows stylesheets to be customized to give
different output for different output formats.  However,
stylesheets customized in this way will not be useable
by other CSL implementations.

# OPTIONS

`-y, --bib2yaml`
:   Convert bibliography to YAML suitable for inclusion in pandoc metadata.

`-j, --bib2json`
:   Convert bibliography to CSL JSON suitable for import into Zotero.

`-f` *FORMAT*, `--format=`*FORMAT*
:   Specify format of bibliography to be converted.  Legal values are
    `biblatex`, `bibtex`, `ris`, `endnote`, `endnotexml`, `isi`,
    `medline`, `copac`, `mods`, `nbib`, and `json`.

`-c` *NUMBER*, `--columns=`*NUMBER*
:   Specify column to which to wrap YAML output produced using
    `--bib2yaml`.  Default is 80.  To disable wrapping, set 0.

`-h, --help`
:   Print usage information.

`--man`
:   Print the man page in groff man format.  To get a plain text
    version, `pandoc-citeproc --man | groff -mman -Tutf8`.  To get an HTML
    version, `pandoc-citeproc --man | groff -mman -Thtml`.

`--license`
:   Print the license.

`-q, --quiet`
:   Silence all warnings.

`-V, --version`
:   Print version.

# NOTES

## General

If you use a biblatex database, closely follow the specifications in the
"Database Guide" section of the biblatex manual (currently 2.8a).

If you use a CSL-YAML or CSL-JSON database, or a CSL-YAML metadata section in
your markdown document, follow the "Citation Style Language 1.0.1 Language
Specification" (<http://citationstyles.org/downloads/specification.html>).
Particularly relevant are
<http://citationstyles.org/downloads/specification.html#appendix-iii-types>
(which neither comments on usage nor specifies required and optional fields) and
<http://citationstyles.org/downloads/specification.html#appendix-iv-variables>
(which does contain comments).

## Titles: Title vs. Sentence Case

If you are using a bibtex or biblatex bibliography, then observe
the following rules:

  - English titles should be in title case.  Non-English titles should
    be in sentence case, and the `langid` field in biblatex should be
    set to the relevant language.  (The following values are treated
    as English:  `american`, `british`, `canadian`, `english`,
    `australian`, `newzealand`, `USenglish`, or `UKenglish`.)

  - As is standard with bibtex/biblatex, proper names should be
    protected with curly braces so that they won't be lowercased
    in styles that call for sentence case.  For example:

        title = {My Dinner with {Andre}}

  - In addition, words that should remain lowercase (or camelCase)
    should be protected:

        title = {Spin Wave Dispersion on the {nm} Scale}

    Though this is not necessary in bibtex/biblatex, it is necessary
    with citeproc, which stores titles internally in sentence case,
    and converts to title case in styles that require it.  Here we
    protect "nm" so that it doesn't get converted to "Nm" at this stage.

If you are using a CSL bibliography (either JSON or YAML), then observe
the following rules:

  - All titles should be in sentence case.

  - Use the `language` field for non-English titles to prevent their
    conversion to title case in styles that call for this. (Conversion
    happens only if `language` begins with `en` or is left empty.)

  - Protect words that should not be converted to title case using
    this syntax:

        Spin wave dispersion on the <span class="nocase">nm</span> scale

## Conference Papers, Published vs. Unpublished

For a formally published conference paper, use the biblatex entry type
`inproceedings` (which will be mapped to CSL `paper-conference`).

For an unpublished manuscript, use the biblatex entry type `unpublished`
without an `eventtitle` field (this entry type will be mapped to CSL
`manuscript`).

For a talk, an unpublished conference paper, or a poster presentation, use the
biblatex entry type `unpublished` with an `eventtitle` field (this entry type
will be mapped to CSL `speech`). Use the biblatex `type` field to indicate the
type, e.g. "Paper", or "Poster". `venue` and `eventdate` may be useful too,
though `eventdate` will not be rendered by most CSL styles. Note that `venue`
is for the event's venue, unlike `location` which describes the publisher's
location; do not use the latter for an unpublished conference paper.

# AUTHORS

Andrea Rossato and John MacFarlane.

# SEE ALSO

`pandoc` (1).

The pandoc-citeproc source code and all documentation may be downloaded
from <http://github.com/jgm/pandoc-citeproc/>.


changelog
pandoc-citeproc (0.16.4)

  * Simplify compat since we now require pandoc >= 2.8.
  * Update man page.  Clarify that `reference-section-title` doesn't have any
    effect when an explicit Div is used (#424).
  * Qualify fail with Prelude to fix compiler warnings.
  * Disable `raw_attribute` extension when writing Markdown.
    This way we avoid e.g. `<i>{=html}` in the CSL export.
  * Use pandoc-types 1.20.  Note that this change removes
    the ability to compile pandoc-citeproc with
    older versions of pandoc (< 2.8).
  * Incorporate the switch to Text in the dependencies (Christian
    Despres).  The changes to the structure of the code are fairly minimal.
    None of the types have changed, requiring a reasonable amount of
    packing, unpacking, and view patterns.

pandoc-citeproc (0.16.3.1)

  * Fix how LANG is set for bibtex conversion.
    We were using a `-` where `_` is standard.
  * Make locale retrieval more robust (#420).
    Previously an error was raised if the locale was 'C'.
    Now 'C' is treated as default (en-US locale is used).

pandoc-citeproc (0.16.3)

  * Set `hanging-indent` class on `refs` Div if specified by the style (#410).
    This depends on whether the `hanging-indent` attribute of
    `<bibliography>` is true.
  * Update chicago-author-date.csl and locales.
  * Add new biblatex @dataset entry type (#413).
  * Cleanup cabal file and remove old ghc 6.10 conditional.
  * Replace some of the yaml use with HsYAML-aeson (vijayphoenix).
  * Allow xml-conduit 1.9.
  * Allow network 3.1.
  * Allow pandoc 2.8.
  * Use pandoc-types 1.17.6.1

pandoc-citeproc (0.16.2)

  * Normalize for sorting by padding numerical strings with 0s (#399).
    This is a somewhat ugly fix to #399, and might be wrong
    if there are strings that may be numerical and should not
    be sorted in numerical order.  But this is the best we can
    do right now without major archictectural changes.
  * Add Ord instance for CNum.  [Minor API change.]
  * Allow empty sort-separator.
    Previously this was overridden with the default.
    Text.CSL.Common: export getOptionValWithDefault [API change].
  * Fix omitted phrase in --man documentation (#394, Daniel Himmelstein).
  * Amend fix to #392 (#393).  Initials were being dropped when there were
    more than two unspaced initials.

pandoc-citeproc (0.16.1.3)

  * Correctly handle initials even when
    presented without a space, as in "J.G. Smith" (#392).
  * Avoid generating Str "" when possible.
  * Allow superscript in-text styles to be treated as "notes"
    for purposes of moving punctuation (George Pollard, #382).

pandoc-citeproc (0.16.1.2)

  * Fix extraction page-first when page is not fully numeric (#386).
  * For cites in notes in a footnote style, don't add comma right after
    open parentheses.  Partially addresses #389.
  * Ensure that URIs (e.g. in DOIs) are properly escaped (#391).

pandoc-citeproc (0.16.1.1)

  * Bump version bounds for pandoc, network.

pandoc-citeproc (0.16.1)

  * Add --columns option.  This allows users to specify the column
    width for text wrapping in --bib2yaml output.
  * Use --columns=0 in biblio2yaml tests to avoid problems with
    different wrapping in different libyaml versions.
  * man page: remove references to obsolete pandoc_markdown(5) man page.
  * man page: add info on how to set metadata (#378).

pandoc-citeproc (0.16.0.2)

  * Remove need for HTTP access in test for #213.

pandoc-citeproc (0.16.0.1)

  * Fix regression on #213.  Literals without special handling
    (such as call-number, dimensions, scale, language) were being
    dropped.  Added regression test.

pandoc-citeproc (0.16)

  * Detect en dash in page range (Andrew Dunning).  Ensure that `--` or `–`
    are detected as page ranges. This fixes a regression that caused the
    second half of the range to be appended to a URL in note styles.

  * Fix collapsing behavior with items with status but no year (#371).

  * Style: Add OStatus constructor to Output [API change].

  * Allow pandoc 2.6.

  * Added pandoc-citeproc to build-tools for test suite (#369).

pandoc-citeproc (0.15.0.1)


  * Allow pandoc 2.5.

pandoc-citeproc (0.15)

  * Support BibLaTeX extended name format (#266).
  * Add refOtherIds field to Reference, and populate it with
    the 'ids' field in biblatex (#356).  Search on refOtherIds
    as well as refId when looking for references.
  * Allow the extent of the locator to be explicitly indicated
    using curly braces, as in:  [@key{67}], [@key, {p. 95 A 3}, suffix]
    (Cormac Relf).
  * Parse balanced brackets in locators like '1(2)(a)(i)' (Cormac Relf).
    This is useful for citing legislation. It is necessary to be more liberal
    in locator parsing because some styles put other elements after a locator,
    like short titles in the case of AGLC, and if you can't parse a locator,
    it gets rendered after, or worse, split in half.
  * Drop ibid-with-locator when prev cite has same locator and label (#241,
    Cormac Relf).
    ```
    [@key]      => "first"
    [@key]      => "ibid"
    [@key, 53]  => "ibid-with-locator"
    [@key, 123] => "ibid-with-locator"
    [@key, 123] => "ibid"
    [@key]      => "subsequent"
    ```
    See http://docs.citationstyles.org/en/stable/specification.html#locators
  * Fix bug wherein nonbreaking spaces classed a name as East Asian,
    so that spaces aren't used between given and family (#365).
    Also regularized logic for detecting "byzantine" names
    (those that get the space).  Now a name is treated as
    byzantine if there are any byzantine characters in it.
    Previously some checks required all byzantine characters.
  * Add mktest.sh for wrangling test cases (#364, Cormac Relf).
  * Update pandoc-citeproc man page.
  * Update locales el-GR, fi-FI.
  * Add CONTRIBUTING.md.

pandoc-citeproc (0.14.8.1)

  * Avoid putting a space after nonbreaking space at end of prefix.
    Previously this resulted in overly long spaces after abbreviations

pandoc-citeproc.hs
{-# LANGUAGE CPP                 #-}
{-# LANGUAGE NoImplicitPrelude   #-}
{-# LANGUAGE OverloadedStrings   #-}
{-# LANGUAGE ScopedTypeVariables #-}
module Main where
import           Prelude
import           Control.Applicative              (many, (<|>))
import           Control.Exception                as E
import           Control.Monad
import           Data.Aeson.Encode.Pretty         (Config (..), Indent (Spaces),
                                                   NumberFormat (Generic),
                                                   defConfig, encodePretty')
import           Data.Attoparsec.ByteString.Char8 as Attoparsec
import qualified Data.ByteString                  as B
import qualified Data.ByteString.Char8            as B8
import qualified Data.ByteString.Lazy             as BL
import           Data.Char                        (chr, toLower)
import           Data.List                        (group, sort)
import qualified Data.Text                        as T
import           Data.Text.Encoding               (encodeUtf8)
import           Data.Version                     (showVersion)
import           Data.Yaml.Builder                (toByteStringWith, setWidth)
import           Text.Libyaml                     (defaultFormatOptions)
import           Paths_pandoc_citeproc            (version)
import           System.Console.GetOpt
import           System.Environment               (getArgs)
import           System.Exit
import           System.FilePath                  (takeExtension)
import           System.IO
import           Text.CSL.Data                    (getLicense, getManPage)
import           Text.CSL.Exception
import           Text.CSL.Input.Bibutils          (BibFormat (..),
                                                   readBiblioString)
import           Text.CSL.Pandoc                  (processCites')
import           Safe                             (readMay)
import           Text.CSL.Reference               (Literal (..),
                                                   Reference (refId))
import           Text.Pandoc.JSON                 hiding (Format)
import qualified Text.Pandoc.UTF8                 as UTF8
import           Text.Pandoc.Walk

main :: IO ()
main = do
  argv <- getArgs
  let (flags, args, errs) = getOpt Permute options argv
  let header = "Usage: pandoc-citeproc [options] [file..]"
  unless (null errs) $ do
    UTF8.hPutStrLn stderr $ usageInfo (unlines $ errs ++ [header]) options
    exitWith $ ExitFailure 1
  when (Version `elem` flags) $ do
    UTF8.putStrLn $ "pandoc-citeproc " ++ showVersion version
    exitSuccess
  when (Help `elem` flags) $ do
    UTF8.putStrLn $ usageInfo header options
    exitSuccess
  when (Man `elem` flags) $ do
    getManPage >>= BL.putStr
    exitSuccess
  when (License `elem` flags) $ do
    getLicense >>= BL.putStr
    exitSuccess

  E.handle
    (\(e :: CiteprocException) -> do
        UTF8.hPutStrLn stderr $ renderError e
        exitWith (ExitFailure 1)) $
    if Bib2YAML `elem` flags || Bib2JSON `elem` flags
       then do
         colwidth <- case Prelude.take 1 $
                          reverse [readMay n | Columns n <- flags] of
                          []        -> return $ Just 80
                          [Just 0]  -> return $ Nothing
                          [Just x]  -> return $ Just x
                          _         -> do
                            UTF8.hPutStrLn stderr $
                              "--columns must be followed by a number"
                            exitWith $ ExitFailure 7
         let fmtOpts = setWidth colwidth defaultFormatOptions
         let mbformat = case [f | Format f <- flags] of
                             [x] -> readFormat x
                             _   -> Nothing
         bibformat <- case mbformat <|>
                           msum (map formatFromExtension args) of
                           Just f   -> return f
                           Nothing  -> do
                              UTF8.hPutStrLn stderr $ usageInfo
                                ("Unknown format\n" ++ header) options
                              exitWith $ ExitFailure 4
         bibstring <- case args of
                           [] -> UTF8.getContents
                           xs -> mconcat <$> mapM UTF8.readFile xs
         readBiblioString (const True) bibformat bibstring >>=
           (if Quiet `elem` flags then return else warnDuplicateKeys) >>=
           if Bib2YAML `elem` flags
              then outputYamlBlock .
                   B8.intercalate (B.singleton 10) .
                   map (unescapeTags . toByteStringWith fmtOpts . (:[]))
              else B8.putStrLn . unescapeUnicode . B.concat . BL.toChunks .
                encodePretty' defConfig{ confIndent = Spaces 2
                                       , confCompare = compare
                                       , confNumFormat = Generic }
       else toJSONFilter (doCites (Quiet `elem` flags))

formatFromExtension :: FilePath -> Maybe BibFormat
formatFromExtension = readFormat . dropWhile (=='.') . takeExtension

readFormat :: String -> Maybe BibFormat
readFormat = go . map toLower
  where go "biblatex"   = Just BibLatex
        go "bib"        = Just BibLatex
        go "bibtex"     = Just Bibtex
        go "json"       = Just Json
        go "yaml"       = Just Yaml
#ifdef USE_BIBUTILS
        go "ris"        = Just Ris
        go "endnote"    = Just Endnote
        go "enl"        = Just Endnote
        go "endnotexml" = Just EndnotXml
        go "xml"        = Just EndnotXml
        go "wos"        = Just Isi
        go "isi"        = Just Isi
        go "medline"    = Just Medline
        go "copac"      = Just Copac
        go "mods"       = Just Mods
        go "nbib"       = Just Nbib
#endif
        go _            = Nothing


doCites :: Bool -> Pandoc -> IO Pandoc
doCites beQuiet doc = do
  doc' <- processCites' doc
  let warnings = query findWarnings doc'
  unless beQuiet $ mapM_ (UTF8.hPutStrLn stderr) warnings
  return doc'

findWarnings :: Inline -> [String]
findWarnings (Span (_,["citeproc-not-found"],[("data-reference-id",ref)]) _) =
  ["pandoc-citeproc: reference " ++ T.unpack ref ++ " not found" | ref /= "*"]
findWarnings (Span (_,["citeproc-no-output"],_) _) =
  ["pandoc-citeproc: reference with no printed form"]
findWarnings _ = []

data Option =
      Help
    | Man
    | License
    | Version
    | Convert
    | Format String
    | Columns String
    | Bib2YAML
    | Bib2JSON
    | Quiet
  deriving (Ord, Eq, Show)

options :: [OptDescr Option]
options =
  [ Option ['h'] ["help"] (NoArg Help) "show usage information"
  , Option [] ["man"] (NoArg Man) "print man page to stdout"
  , Option [] ["license"] (NoArg License) "print license to stdout"
  , Option ['V'] ["version"] (NoArg Version) "show program version"
  , Option ['y'] ["bib2yaml"] (NoArg Bib2YAML) "convert bibliography to YAML"
  , Option ['j'] ["bib2json"] (NoArg Bib2JSON) "convert bibliography to JSON"
  , Option ['q'] ["quiet"] (NoArg Quiet) "silence all warnings"
  , Option ['f'] ["format"] (ReqArg Format "FORMAT") "bibliography format"
  , Option ['c'] ["columns"] (ReqArg Columns "NUMBER") "column width (or 0)"
  ]

warnDuplicateKeys :: [Reference] -> IO [Reference]
warnDuplicateKeys refs = mapM_ warnDup dupKeys >> return refs
  where warnDup k = UTF8.hPutStrLn stderr $ "biblio2yaml: duplicate key " ++ k
        allKeys   = map (unLiteral . refId) refs
        dupKeys   = [x | (x:_:_) <- group (sort allKeys)]

outputYamlBlock :: B.ByteString -> IO ()
outputYamlBlock contents = do
  UTF8.putStrLn "---\nreferences:"
  B.putStr contents
  UTF8.putStrLn "..."

-- turn
-- id: ! "\u043F\u0443\u043D\u043A\u04423"
-- into
-- id: пункт3
unescapeTags :: B.ByteString -> B.ByteString
unescapeTags bs = case parseOnly (many $ tag <|> other) bs of
                       Left e  -> error e
                       Right r -> B.concat r

unescapeUnicode :: B.ByteString -> B.ByteString
unescapeUnicode bs = case parseOnly (many other) bs of
                       Left e  -> error e
                       Right r -> B.concat r

tag :: Attoparsec.Parser B.ByteString
tag = do
  _ <- string $ B8.pack ": ! "
  c <- char '\'' <|> char '"'
  cs <- manyTill (escaped c <|> other) (char c)
  return $ B8.pack ": " <> B8.singleton c <> B.concat cs <> B8.singleton c

escaped :: Char -> Attoparsec.Parser B.ByteString
escaped c = string $ B8.pack ['\\',c]

other :: Attoparsec.Parser B.ByteString
other = uchar <|> Attoparsec.takeWhile1 notspecial <|> regchar
  where notspecial = not . inClass ":!\\\"'"

uchar :: Attoparsec.Parser B.ByteString
uchar = do
  _ <- char '\\'
  num <- (2 <$ char 'x') <|> (4 <$ char 'u') <|> (8 <$ char 'U')
  cs <- count num $ satisfy $ inClass "0-9a-fA-F"
  let n = read ('0':'x':cs)
  return $ encodeUtf8 $ T.pack [chr n]

regchar :: Attoparsec.Parser B.ByteString
regchar = B8.singleton <$> anyChar

